apiserver/common/leadership.go

```diff
@@ -103,47 +107,108 @@ func (a *leadershipPinningAPI) PinnedLeadership() (params.PinnedLeadershipResult
 }
 
-// pinMachineAppsOps runs the input pin/unpin operation against all
-// applications represented by units on the authorised machine.
-// An assumption is made that the validity of the auth tag has been verified
-// by the caller.
-func (a *leadershipPinningAPI) pinMachineAppsOps(op func(string, string) error) (params.PinApplicationsResults, error) {
-	result := params.PinApplicationsResults{}
+// GetMachineApplicationNames returns the applications associated with a
+// machine.
+func (a *LeadershipPinning) GetMachineApplicationNames() ([]string, error) {
+	if !a.authorizer.AuthMachineAgent() {
+		return nil, ErrPerm
+	}
 
 	tag := a.authorizer.GetAuthTag()
 	m, err := a.st.Machine(tag.Id())
 	if err != nil {
-		return result, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 	apps, err := m.ApplicationNames()
 	if err != nil {
-		return result, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
+	return apps, nil
+}
+
+// PinMachineApplicationsByName takes a slice of application names and attempts
+// to pin them accordingly.
+func (a *LeadershipPinning) PinMachineApplicationsByName(appNames []string) (params.PinApplicationsResults, error) {
+	if !a.authorizer.AuthMachineAgent() {
+		return params.PinApplicationsResults{}, ErrPerm
+	}
+
+	return a.pinMachineAppsOps(appNames, a.pinner.PinLeadership)
+}
+
+// UnpinMachineApplicationsByName takes a slice of application names and
+// attempts to pin them accordingly.
+func (a *LeadershipPinning) UnpinMachineApplicationsByName(appNames []string) (params.PinApplicationsResults, error) {
+	if !a.authorizer.AuthMachineAgent() {
+		return params.PinApplicationsResults{}, ErrPerm
+	}
+
+	return a.pinMachineAppsOps(appNames, a.pinner.UnpinLeadership)
+}
+
+// pinMachineApplications pins leadership for applications represented by units
+// running on the auth'd machine.
+func (a *LeadershipPinning) pinMachineApplications() (params.PinApplicationsResults, error) {
+	appNames, err := a.GetMachineApplicationNames()
+	if err != nil {
+		return params.PinApplicationsResults{}, ErrPerm
+	}
+	return a.pinMachineAppsOps(appNames, a.pinner.PinLeadership)
+}
+
+// unpinMachineApplications unpins leadership for applications represented by
+// units running on the auth'd machine.
+func (a *LeadershipPinning) unpinMachineApplications() (params.PinApplicationsResults, error) {
+	appNames, err := a.GetMachineApplicationNames()
+	if err != nil {
+		return params.PinApplicationsResults{}, ErrPerm
+	}
+	return a.pinMachineAppsOps(appNames, a.pinner.UnpinLeadership)
+}
+
+// pinMachineAppsOps runs the input pin/unpin operation against all
+// applications represented by units on the authorised machine.
+// An assumption is made that the validity of the auth tag has been verified
+// by the caller.
+func (a *LeadershipPinning) pinMachineAppsOps(appNames []string, op func(string, string) error) (params.PinApplicationsResults, error) {
+	var result params.PinApplicationsResults
+
+	tag := a.authorizer.GetAuthTag()
```

Comment:
This isn't going to work how we want it to for the `remove-machine` case, because we will auth as a client.
This tag needs to be passed in from higher up.